### PR TITLE
Add SRem and URem visit

### PR DESCRIPTION
**Pull request by Skrinikov**
Not sure if more tests are needed.

I was too lazy to test division by 0.

### CODE REVIEW

**Comment by Skrinikov** [comment]
Fixes #22

File: src/decaf.cpp [hunk 1]
```diff
@@ -278,6 +278,39 @@ ExecutionResult Interpreter::visitUDiv(llvm::BinaryOperator &op) {
...
+
+  auto lhs = normalize_to_int(frame.lookup(op.getOperand(0), *z3));
+  auto rhs = normalize_to_int(frame.lookup(op.getOperand(1), *z3));
+
+  if (ctx->check(rhs == 0) == z3::sat) {
+    queue->add_failure(ctx->solver.get_model());
+  }
+  ctx->add(rhs != 0);
+
+  frame.insert(&op, lhs % rhs);
```
Comment (Phantomical) [inline src/decaf.cpp:309]: Signed and unsigned remainder are different so you can't use the `%` operator for both cases. You should use the operation for unsigned remainder here. It's `z3::urem`.

It looks like the division operator has the same problem, if you could replace the unsigned division with `z3::udiv` that would be best.

File: tests/opcodes.cpp [hunk 2]
```diff
@@ -283,3 +283,77 @@ TEST(opcodes, sdiv_test_overflow) {
...
+  auto &bb = func->getEntryBlock();
+  auto div1 = llvm::BinaryOperator::CreateSRem(val1, val2, "div1", &bb);
+
+  interp.visitSRem(*div1);
+
+  // For now there should be a print msg in cout that we tried to divide by 0.
+  // TODO: Change this test once `Executor::add_failure` is updated.
+  auto expr0 = ctx.stack_top().lookup(div1, z3);
+  EXPECT_EQ(ctx.check(expr0 == 2147483647), z3::unsat);
+}
```
Comment (Phantomical) [inline tests/opcodes.cpp:359]: Maybe add a test case for successful remainder by a negative number? (e.g. 20 % -4)

**Review by Phantomical — CHANGES_REQUESTED** [review]
The incorrect remainder operation is my main concern. Fix that and I'll approve.

The extra test case is something that we should add (it would have caught the bug) but I won't block approval on it.

File: tests/opcodes.cpp [hunk 3]
```diff
@@ -283,3 +283,77 @@ TEST(opcodes, sdiv_test_overflow) {
...
+
+  Context ctx{z3, func.get()};
+  Interpreter interp{&ctx, nullptr, &z3};
+
+  auto &bb = func->getEntryBlock();
+  auto div1 = llvm::BinaryOperator::CreateSRem(val1, val2, "div1", &bb);
+
+  interp.visitSRem(*div1);
+
+  // For now there should be a print msg in cout that we tried to divide by 0.
```
Comment (mishazharov) [inline tests/opcodes.cpp:355]: nit: is this comment accurate? I think some copy pasta happened here

**Review by Phantomical — APPROVED** [review]
LGTM! Here's a free test case you might want to add to run-pass

```cpp
#include "decaf.h"
#include <stdint.h>

void test(int32_t x) {
  decaf_assume(x == -50);
  int32_t y = x / -5;
  decaf_assert(y == 10);
}
```

Also `queue->add_failure` is now `tracker->add_failure`. You'll need to fix this.
